onchain/solidity_contracts/src/defi/ABTCVault.sol

```diff
@@ -0,0 +1,112 @@
+
+
+contract ABTCVault is Initializable, ERC20Upgradeable, ERC4626Upgradeable, AccessControlUpgradeable, UUPSUpgradeable, PausableUpgradeable {
+    function decimals() public view virtual override(ERC20Upgradeable, ERC4626Upgradeable) returns (uint8) {
+        return 18;
+    }
+    bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+    bytes32 public constant UPGRADER_ROLE = keccak256("UPGRADER_ROLE");
+    bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+
+    struct WrappedBTC {
+        bool isPermitted;
+        uint256 ratio; 
+        uint256 poolingTimestamp;
+    }
+
+    mapping(address => WrappedBTC) public wrappedBTCTokens;
+    mapping(address => mapping(address => uint256)) public userDeposits;
+
+    event TokenPermissionSet(address indexed token, bool isPermitted, uint256 ratio, uint256 poolingTimestamp);
+    event Deposited(address indexed user, address indexed token, uint256 amount, uint256 mintedAmount);
+    event Withdrawn(address indexed user, address indexed token, uint256 amount, uint256 burnedAmount);
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(address _admin, address _underlying) initializer public {
+        __ERC20_init("Aggregated Bitcoin", "aBTC");
+        __ERC4626_init(IERC20(_underlying));
+        __AccessControl_init();
+        __UUPSUpgradeable_init();
+        __Pausable_init();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, _admin);
+        _grantRole(MINTER_ROLE, _admin);
+        _grantRole(UPGRADER_ROLE, _admin);
+        _grantRole(PAUSER_ROLE, _admin);
+    }
+
+    function setWrappedBTCToken(address _token, bool _isPermitted, uint256 _ratio, uint256 _poolingTimestamp) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(_ratio > 0, "Invalid ratio");
+        wrappedBTCTokens[_token] = WrappedBTC(_isPermitted, _ratio, _poolingTimestamp);
+        emit TokenPermissionSet(_token, _isPermitted, _ratio, _poolingTimestamp);
+    }
+
+    function deposit(address _wrappedBTCToken, uint256 _amount) external whenNotPaused {
+        WrappedBTC memory wrappedToken = wrappedBTCTokens[_wrappedBTCToken];
+        require(wrappedToken.isPermitted, "Token not permitted");
+        
+        uint256 aBTCAmount = (_amount * wrappedToken.ratio) / 1e18;
+        
+        IERC20(_wrappedBTCToken).transferFrom(msg.sender, address(this), _amount);
+        _mint(msg.sender, aBTCAmount);
+
+        userDeposits[msg.sender][_wrappedBTCToken] += _amount;
+        
+        emit Deposited(msg.sender, _wrappedBTCToken, _amount, aBTCAmount);
+    }
+
+    function withdraw(address _wrappedBTCToken, uint256 _aBTCAmount) external whenNotPaused {
+        WrappedBTC memory wrappedToken = wrappedBTCTokens[_wrappedBTCToken];
+        require(wrappedToken.isPermitted, "Token not permitted");
+        
+        uint256 wrappedBTCAmount = (_aBTCAmount * 1e18) / wrappedToken.ratio;
```

Comment:
Can we change it without ratio atm please?
We are gonna add a fees percentage after as a ratio for a protocol fee.
Can be setup to 0 to do the test, or just comment it